backend/corpora/common/entities/dataset_asset.py

```diff
@@ -0,0 +1,53 @@
+import logging
+import typing
+
+import boto3
+from botocore.exceptions import ClientError
+
+from .entity import Entity
+from ..corpora_orm import DbDatasetArtifact
+from urllib.parse import urlparse
+
+logger = logging.getLogger(__name__)
+
+
+class DatasetAsset(Entity):
+    table = DbDatasetArtifact
+    s3 = boto3.client("s3")
+
+    def __init__(self, db_object: DbDatasetArtifact):
+        super().__init__(db_object)
+
+        url = urlparse(self.s3_uri)
+        self.bucket_name = url.netloc
+        self.key_name = url.path[1:]
+
+    def generate_file_url(self, expiration: int = 604800) -> typing.Union[str, None]:
+        """
+        Generate a presigned URL for a file for user download.
+        :param expiration: Presigned URL expiration in seconds. The default is 1 week.
+        :return: Presigned URL to download the requested file
+        """
+        try:
+            response = self.s3.generate_presigned_url(
+                "get_object", Params={"Bucket": self.bucket_name, "Key": self.key_name}, ExpiresIn=expiration
+            )
+        except ClientError:
+            logger.exception(f"Failed to generate presigned URL for '{self.key_name}'.")
+            return None
+        else:
+            return response
+
+    def get_file_size(self) -> typing.Union[int, None]:
+        """
+        Retrieves the asset content length from the S3 object.
+        :return: The content length in bytes.
+        """
+
+        try:
+            response = self.s3.head_object(Bucket=self.bucket_name, Key=self.key_name)
+        except ClientError:
+            logger.exception(f"Failed to retrieve meta data for '{self.key_name}'.")
```

Comment:
Nit: I'd maybe consider logging a more helpful error message since it'll be read by us. Something like `f"Could not locate S3 object s3://{self.bucket_name}/{self.key_name} to fetch metadata."`. Or something like that.

Reply:
I add the original URL to the response.